core/src/main/java/org/apache/calcite/rel/rules/SingleValueOptimizationRules.java

```diff
@@ -0,0 +1,329 @@
+              join.getJoinType())).transform(relBuilder);
+      if (transformed != null) {
+        call.transformTo(transformed);
+      }
+    }
+
+    static Predicate<Join> elibilityPredicate(boolean isLeft) {
+      if (isLeft) {
+        return jn -> jn.getJoinType() == JoinRelType.LEFT
+            || jn.getJoinType() == JoinRelType.FULL
+            || jn.getJoinType() == JoinRelType.ANTI;
+      } else {
+        return jn -> jn.getJoinType() == JoinRelType.RIGHT
+            || jn.getJoinType() == JoinRelType.FULL
+            || jn.getJoinType() == JoinRelType.ANTI;
+      }
+    }
+
+    @Override public boolean autoPruneOld() {
+      return true;
+    }
+
+    /** Rule configuration. */
+    protected interface Config extends RelRule.Config {
+      @Override PruneSingleValueRule toRule();
+    }
+  }
+
+  /** Configuration for a rule that simplifies join node with constant row on its right input. */
+  @Value.Immutable
+  interface JoinRightSingleRuleConfig extends PruneSingleValueRule.Config {
+    JoinRightSingleRuleConfig DEFAULT = ImmutableJoinRightSingleRuleConfig.of()
+        .withOperandSupplier(b0 ->
+            b0.operand(Join.class).inputs(
+                b1 -> b1.operand(RelNode.class).anyInputs(),
+                b2 -> b2.operand(Values.class).predicate(Values::isSingleValue).noInputs()))
+        .withDescription("PruneJoinSingleValue(right)");
+
+    @Override default SingleValueOptimizationRules.PruneSingleValueRule toRule() {
+      return new SingleValueOptimizationRules.PruneSingleValueRule(this) {
+        @Override public void onMatch(RelOptRuleCall call) {
+          final Join join = call.rel(0);
+          final Values values = call.rel(2);
+          final RelNode other = call.rel(1);
+          onMatch(call, values, null, join, other, false);
+        }
+      };
+    }
+  }
+
+  /** Configuration for a rule that simplifies join node with constant row on its left input. */
+  @Value.Immutable
+  interface JoinLeftSingleRuleConfig extends PruneSingleValueRule.Config {
+    JoinLeftSingleRuleConfig DEFAULT = ImmutableJoinLeftSingleRuleConfig.of()
+        .withOperandSupplier(b0 ->
+            b0.operand(Join.class).inputs(
+                b1 -> b1.operand(Values.class).predicate(Values::isSingleValue).noInputs(),
+                b2 -> b2.operand(RelNode.class).anyInputs()))
+        .withDescription("PruneJoinSingleValueRule(left)");
+
+    @Override default SingleValueOptimizationRules.PruneSingleValueRule toRule() {
+      return new SingleValueOptimizationRules.PruneSingleValueRule(this) {
+        @Override public void onMatch(RelOptRuleCall call) {
+          final Join join = call.rel(0);
+          final Values values = call.rel(1);
+          final RelNode other = call.rel(2);
+          onMatch(call, values, null, join, other, true);
+        }
+      };
+    }
+  }
+
+  /** Configuration for a rule that simplifies join node with a project on a constant row
```

Comment:
what is a project on a constant row? Is it a projection applied to a VALUES node?
Isn't that case handled by a separate optimization step?

Reply:
This should be a comment in the code.

Reply:
please add this as a comment in the code too.